core/src/main/java/org/elasticsearch/action/admin/indices/shrink/TransportShrinkAction.java

```diff
@@ -0,0 +1,210 @@
+                        listener.onResponse(new ShrinkResponse(response.isAcknowledged()));
+                    }
+
+                    @Override
+                    public void onFailure(Throwable t) {
+                        if (t instanceof IndexAlreadyExistsException) {
+                            logger.trace("[{}] failed to create shrink index", t, updateRequest.index());
+                        } else {
+                            logger.debug("[{}] failed to create shrink index", t, updateRequest.index());
+                        }
+                        listener.onFailure(t);
+                    }
+                });
+            }
+
+            @Override
+            public void onFailure(Throwable e) {
+                listener.onFailure(e);
+            }
+        });
+
+    }
+
+    // static for unittesting this method
+    static CreateIndexClusterStateUpdateRequest prepareCreateIndexRequest(final ShrinkRequest shrinkReqeust, final ClusterState state
+        , final DocsStats docsStats, IndexNameExpressionResolver indexNameExpressionResolver) {
+        final String sourceIndex = indexNameExpressionResolver.resolveDateMathExpression(shrinkReqeust.getSourceIndex());
+        final CreateIndexRequest targetIndex = shrinkReqeust.getShrinkIndexReqeust();
+        final String targetIndexName = indexNameExpressionResolver.resolveDateMathExpression(targetIndex.index());
+        final IndexMetaData metaData = state.metaData().index(sourceIndex);
+        if (metaData == null) {
+            throw new IndexNotFoundException(sourceIndex);
+        }
+        // ensure index is read-only
+        if (state.blocks().indexBlocked(ClusterBlockLevel.WRITE, sourceIndex) == false) {
+            throw new IllegalStateException("index " + sourceIndex + " must be read-only to shrink index. use \"index.blocks.write=true\"");
+        }
+        if (state.metaData().hasIndex(targetIndexName)) {
+            throw new IndexAlreadyExistsException(state.metaData().index(targetIndexName).getIndex());
+        }
+        if (metaData.getNumberOfShards() == 1) {
+            throw new IllegalArgumentException("can't shrink an index with only one shard");
+        }
+        long count = docsStats.getCount();
+        if (count >= IndexWriter.MAX_DOCS) {
+            throw new IllegalStateException("Can't merge index with more than [" + IndexWriter.MAX_DOCS
+                + "] docs -  too many documents");
+        }
+        if (targetIndex.mappings().isEmpty() == false) {
+            throw new IllegalArgumentException("mappings are not allowed when shrinking indices" +
+                ", all mappings are copied from the source index");
+        }
+        final Settings tragetIndexSettings = Settings.builder().put(targetIndex.settings())
+            .normalizePrefix(IndexMetaData.INDEX_SETTING_PREFIX).build();
+        if (IndexMetaData.INDEX_NUMBER_OF_SHARDS_SETTING.exists(tragetIndexSettings)
```

Comment:
this validation is duplicated in the MetaDataCreateIndexService including some more checks regarding the source index. I think we should just have one place - either here or there. I think both works.